src/main.js

```diff
@@ -0,0 +1,208 @@
+  constructor(opts = {}) {
+    const settings = Object.assign({}, DEFAULT_OPTIONS);
+    this.settings = Object.assign(settings, opts);
+    // a cache to store memoized srcset width-pairs
+    this.targetWidthsCache = {};
+    if (typeof this.settings.domain != "string") {
+      throw new Error('ImgixClient must be passed a valid string domain');
+    }
+
+    if (DOMAIN_REGEX.exec(this.settings.domain) == null) {
+      throw new Error(
+        'Domain must be passed in as fully-qualified ' +
+        'domain name and should not include a protocol or any path ' +
+        'element, i.e. "example.imgix.net".');
+    }
+
+    if (this.settings.includeLibraryParam) {
+      this.settings.libraryParam = "js-" + VERSION;
+    }
+
+    this.settings.urlPrefix = this.settings.useHTTPS ? 'https://' : 'http://';
+  }
+
+  buildURL(path = '', params = {}) {
+    const sanitizedPath = this._sanitizePath(path);
+
+    const queryParams = this._buildParams(params);
+    if (!!this.settings.secureURLToken) {
+      const securedParams = this._signParams(sanitizedPath, queryParams);
+      return this.settings.urlPrefix + this.settings.domain + sanitizedPath + securedParams;
+    }
+    return this.settings.urlPrefix + this.settings.domain + sanitizedPath + queryParams;
+  }
+
+  _buildParams(params = {}) {
+    const queryParams = [];
+    if (this.settings.libraryParam) {
+      queryParams.push('ixlib=' + this.settings.libraryParam)
+    }
+
+    for (const key in params) {
+      const val = params[key];
+      const encodedKey = encodeURIComponent(key);
+
+      let encodedVal;
+      if (key.substr(-2) === '64') {
+        encodedVal = Base64.encodeURI(val);
+      } else {
+        encodedVal = encodeURIComponent(val);
+      }
+      queryParams.push(encodedKey + "=" + encodedVal);
+    }
+
+    if (queryParams[0]) {
+      queryParams[0] = "?" + queryParams[0];
+    }
+    return queryParams.join('&');
+  }
+
+  _signParams(path, queryParams) {
+    const signatureBase = this.settings.secureURLToken + path + queryParams;
+    const signature = md5(signatureBase);
+
+    if (queryParams.length > 0) {
+      return queryParams = queryParams + "&s=" + signature;
+    } else {
+      return queryParams = "?s=" + signature;
+    }
+  }
+
+  _sanitizePath(path) {
+    // Strip leading slash first (we'll re-add after encoding)
+    let _path = path.replace(/^\//, '');
+
+    if (/^https?:\/\//.test(_path)) {
+      // Use de/encodeURIComponent to ensure *all* characters are handled,
+      // since it's being used as a path
+      _path = encodeURIComponent(_path);
+    } else {
+      // Use de/encodeURI if we think the path is just a path,
+      // so it leaves legal characters like '/' and '@' alone
+      _path = encodeURI(_path).replace(/[#?:]/g, encodeURIComponent);
+    }
+
+    return '/' + _path;
+  }
+
+  buildSrcSet(path, params, options) {
+    const _options = options || {};
+    const _params = params || {};
```

Comment:
We can rewrite these (and any other assignments) to use [newer operators](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators). I mean if we can, why not use them ¯\\_(ツ)_/¯
Something like:
```suggestion
    const _options ||= options;
    const _params ||= params;
```

Reply:
>- I think we should move towards parameterizing the tests. I know it's out of scope here, but just wanted to add that 2¢ somewhere before I forget.

👍 

>- In addition to the TS definitions, is there any reason not to have the main file as `.ts` at this point? At least that way, we can benefit from using TS when developing.

I'm so ready to TypeScript it up :)

>- This is super nitpicky but I can't help but notice that some files have a newline at the end of the file while others don't. We should make that consistent, and configure prettier to enforce it in the future.

Code style is very important (not nitpicky at all). I want to get a few more things done here and then configure our tools to format things exactly how we want them.